File: 백준/Silver/14940. 쉬운 최단거리/쉬운 최단거리.js
const input = require('fs').readFileSync(0).toString().trim().split('\n');
const [n, m] = input.shift().split(' ').map(Number);
const grid = [];
for (let i = 0; i < n; i++) {
    grid.push(input[i].split(' ').map(Number));
}

const dist = [];
for (let i = 0; i < n; i++) {
    const row = [];
    for (let j = 0; j < m; j++) {
        row.push(grid[i][j] === 0 ? 0 : -1);
    }
    dist.push(row);
}

const dx = [-1, 1, 0, 0];
const dy = [0, 0, -1, 1];

function bfs(x, y) {
    const queue = [[x, y]];
    dist[x][y] = 0;

    while (queue.length > 0) {
        const [cx, cy] = queue.shift();
        
        for (let d = 0; d < 4; d++) {
            const nx = cx + dx[d];
            const ny = cy + dy[d];
            
            if ((nx >= 0 && nx < n) && (ny >= 0 && ny < m)) {
                if (dist[nx][ny] === -1 && grid[nx][ny] !== 0) {
                    dist[nx][ny] = dist[cx][cy] + 1;
                    queue.push([nx, ny]);
                }
            }
        }
    }
}

for (let i = 0; i < n; i++) {
    for (let j = 0; j < m; j++) {
        if (grid[i][j] === 2) {
            bfs(i, j);
        }
    }
}

for (let row of dist) {
    console.log(row.join(' '));
}

File: 백준/Silver/14940. 쉬운 최단거리/쉬운 최단거리.py
from collections import deque

n, m = map(int, input().split())
grid = [list(map(int, input().split())) for _ in range(n)]

dist = [[0 if grid[i][j] == 0 else -1 for j in range(m)] for i in range(n)]

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

def bfs(start_x, start_y):
    queue = deque()
    queue.append((start_x, start_y))
    dist[start_x][start_y] = 0

    while queue:
        x, y = queue.popleft()
        
        for d in range(4):
            nx = x + dx[d]
            ny = y + dy[d]
            
            if 0 <= nx < n and 0 <= ny < m:
                if dist[nx][ny] == -1 and grid[nx][ny] != 0:
                    dist[nx][ny] = dist[x][y] + 1
                    queue.append((nx, ny))

for i in range(n):
    for j in range(m):
        if grid[i][j] == 2:
            bfs(i, j)
            
for row in dist:
    print(' '.join(map(str, row)))